app/knights.py

```diff
@@ -0,0 +1,50 @@
+from __future__ import annotations
+
+
+class Knights:
+    def __init__(
+            self,
+            name: str,
+            power: int,
+            hp: int,
+            armour: list[dict],
+            weapon: dict,
+            potion: dict | None,
+            protection: int = 0
+    ) -> None:
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.armour = armour
+        self.weapon = weapon
+        self.potion = potion
+        self.protection = protection
+
+    def arm_himself(self) -> None:
+        self.power += self.weapon["power"]
+
+    def putting_on(self) -> None:
+        for element in self.armour:
+            self.protection += element["protection"]
+
+    def application(self) -> None:
+        if self.potion is not None:
+            if "power" in self.potion["effect"]:
+                self.power += self.potion["effect"]["power"]
+
+            if "protection" in self.potion["effect"]:
+                self.protection += self.potion["effect"]["protection"]
+
+            if "hp" in self.potion["effect"]:
+                self.hp += self.potion["effect"]["hp"]
+
+    def battle(self, other: Knights) -> None:

```

Comment:
The `battle` method directly modifies the `hp` attribute based on the opponent's power and protection. Consider adding checks to ensure that the resulting `hp` does not become negative, as this is handled later but could be more efficient if done immediately.